brainiak/matnormal/matnormal_likelihoods.py

```diff
@@ -0,0 +1,234 @@
+                          'iqf logdet')
+
+    # A' Sigma^{-1}
+    Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+    # (Qinv + A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+    prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+
+    solve = tf.matmul(sigma.Sigma_inv_x(scaled_I(1.0, sigma.size) -
+                      tf.matmul(A, prod_term)), x)
+
+    return solve, logdet
+
+
+def solve_det_conditional(x, sigma, A, Q):
+    """
+    Use matrix inversion lemma for the solve:
+    .. math::
+    (\Sigma - AQ^{-1}A')^{-1} X =\\
+    \Sigma^{-1} + \Sigma^{-1} A (Q -
+    A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1} X
+
+    Use matrix determinant lemma for determinant:
+    .. math::
+    \log|(\Sigma - AQ^{-1}A')| =
+    \log|Q - A' \Sigma^{-1} A| - \log|Q| + \log|\Sigma|
+    """
+
+    # (Q - A' Sigma^{-1} A)
+    i_qf_cholesky = tf.cholesky(Q.Sigma - tf.matmul(A,
+                                sigma.Sigma_inv_x(A), transpose_a=True))
+
+    logdet = -Q.logdet + sigma.logdet +\
+        2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+    # A' Sigma^{-1}
+    Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+    # (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+    prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+
+    solve = tf.matmul(sigma.Sigma_inv_x(scaled_I(1.0, sigma.size) +
+                      tf.matmul(A, prod_term)), x)
+
+    return solve, logdet
+
+
+def _mnorm_logp_internal(colsize, rowsize, logdet_row, logdet_col,
+                         solve_row, solve_col):
+    """Construct logp from the solves and determinants.
+    """
+    log2pi = 1.8378770664093453
+
+    if logging.getLogger().isEnabledFor(logging.DEBUG):
+        solve_row = tf.Print(solve_row, [tf.trace(solve_col)], 'coltrace')
+        solve_row = tf.Print(solve_row, [tf.trace(solve_row)], 'rowtrace')
+        solve_row = tf.Print(solve_row, [logdet_row], 'logdet_row')
+        solve_row = tf.Print(solve_row, [logdet_col], 'logdet_col')
+
+    denominator = - rowsize * colsize * log2pi -\
+        colsize * logdet_row - rowsize * logdet_col
+    numerator = - tf.trace(tf.matmul(solve_col, solve_row))
+    return 0.5 * (numerator + denominator)
+
+
+def matnorm_logp(x, row_cov, col_cov):
+    """Log likelihood for centered matrix-variate normal density.
+    Assumes that row_cov and col_cov follow the API defined in CovBase.
+    """
+
+    rowsize = tf.cast(tf.shape(x)[0], 'float64')
+    colsize = tf.cast(tf.shape(x)[1], 'float64')
+
+    # precompute sigma_col^{-1} * x'
+    solve_col = col_cov.Sigma_inv_x(tf.transpose(x))
+    logdet_col = col_cov.logdet
+
+    # precompute sigma_row^{-1} * x
+    solve_row = row_cov.Sigma_inv_x(x)
+    logdet_row = row_cov.logdet
+
+    return _mnorm_logp_internal(colsize, rowsize, logdet_row,
+                                logdet_col, solve_row, solve_col)
+
+
+def matnorm_logp_marginal_row(x, row_cov, col_cov, marg, marg_cov):
```

Comment:
I don't know the math behind the following functions. But the structure of the codes seem fine.

Reply:
added.